### ISSUE TITLE
introduce `stage` abstraction to testground

### ISSUE THREAD

**Issue by nonsense** [ipfs/testground#875]
All current testplans could use the abstraction of a `stage` that @aschmahmann presented during one of the last Testground calls.

This would also give a bit more visibility over the `progress` of a given testplan to the testground client/daemon and user of Testground.

We should think about:
- an API that testplans use for the stages they init
- a service which queries the sync service / Redis, and finds out the state of the stages that are part of the testplan, and reports back to the testground daemon.

**Comment by nonsense** [comment]
Note that this should replace the stage we currently detect on the `testground daemon` side, namely the "all networks initialised".

**Comment by raulk** [comment]
Some thoughts on this API.

```go
package foo

import (
    "context"

    "github.com/testground/sdk-go/runtime"
    "github.com/testground/sdk-go/stager"
    "github.com/testground/sdk-go/sync"
)

func main() {
    runtime.Invoke(runTest)
}

func runTest(runenv *runtime.RunEnv) error {
    ctx, cancel := context.WithTimeout(context.Background(), 5 * time.Minute)
	defer cancel()

    s := stager.NewSequential(ctx, client)
    s.AddStage("stage1", stage1)
    s.AddStage("stage2", stage2)
    s.AddStage("stage3", stage3)
    return s.Run()
}

func stage1(runenv *runtime.RunEnv) error {
	return nil
}

func stage2(runenv *runtime.RunEnv) error {
	return nil
}

func stage3(runenv *runtime.RunEnv) error {
	return nil
}
```

* Package `stager` can offer more complex stagers going forward, e.g. parallel stagers, conditional stages, etc.
* In this sequential stager, we wait for all test call `client.SignalAndWait()` on a "start" state, and an "end" state before advancing to the next stage.
* If a stage fails, we short-circuit and abort early.

**Comment by coryschwartz** [comment]
With regard to point 1, This is conceptually similar to https://github.com/testground/testground/pull/567 which had a node factory for stages that could be combined into patterns, -- loops, failure paths, sequential operations, etc. Maybe some of that PR useful, even if not using the state machine idea I was going for there.

For part 2, maybe the testground daemon can do this directly. The testground daemon, when running in kubernetes cluster, or executing local runners will have direct access to redis. The only case in which the daemon doesn't have network access to redis is for the k8s runner when the daemon is not part of the cluster, we can tunnel, similarly to the way port forwarding is done: https://github.com/kubernetes/client-go/blob/master/tools/portforward/portforward.go

Does it make sense for the daemon to talk directly to redis without the additional stage query service?

**Comment by nonsense** [comment]
> For part 2, maybe the testground daemon can do this directly. Does it make sense for the daemon to talk directly to redis without the additional stage query service?

I am thinking along the same lines. If we add a few `read` APIs on the `sdk-go` (which essentially is the interface to the `sync service`), then `testground` could call the `sync service` directly, and we don't need an additional `stage query service`.

It would be good for `testground` to use an API provided by `sdk-go`, so that we hide Redis behind the sync service API.

**Comment by raulk** [comment]
Moving this to v0.6. We need to split it in two parts:

- stages API producing diagnostics.
- some facility to monitor the progress of test plans in real-time.

---

@nonsense the issue with making the daemon access Redis directly is that daemon doesn't necessarily have access to Redis, e.g. if running it outside kubernetes, or locally with the docker runner.

**Comment by nonsense** [comment]
> @nonsense the issue with making the daemon access Redis directly is that daemon doesn't necessarily have access to Redis, e.g. if running it outside kubernetes, or locally with the docker runner.

Why not? Redis could still be available and accessible to the docker runner and the local:exec runner... I think a thin API on the sdk-go would be better, than having to deploy another service in the middle.

**Comment by Robmat05** [comment]
@raulk to build wireframes for this which unblocks #599
